### PR TITLE
Upgrade to Ubuntu 22.04 build container

### PR DESCRIPTION
**Pull request by timja**
Downstream of:

- https://github.com/jenkins-infra/docker-packaging/pull/33
- https://github.com/jenkins-infra/docker-packaging/pull/34
- https://github.com/jenkins-infra/docker-packaging/pull/37

see https://github.com/jenkins-infra/release/issues/202

### CODE REVIEW

**Comment by timja** [comment]
`rpm --eval '%{_javadir}'` shows the `_javadir` is no longer defined, investigating.

**Comment by timja** [comment]
Before:

```
jenkins@445d775c2f41:/srv/releases/jenkins$ rpm --eval '%{_javadir}'
/usr/share/java
jenkins@445d775c2f41:/srv/releases/jenkins$  rpm --showrc | grep java
  %{?_javaclasspath:CLASSPATH="%{_javaclasspath}"
-13: __docdir_path	%{_datadir}/doc:%{_datadir}/man:%{_datadir}/info:%{_datadir}/gtk-doc/html:%{?_docdir}:%{?_mandir}:%{?_infodir}:%{?_javadocdir}:/usr/doc:/usr/man:/usr/info:/usr/X11R6/man
-13: _javadir	%{_datadir}/java
-13: _javadocdir	%{_datadir}/javadoc
jenkins@445d775c2f41:/srv/releases/jenkins$ rpm --version
RPM version 4.14.1
```

After:
```
jenkins@64f58e44d47d:/srv/releases/jenkins$ rpm --eval '%{_javadir}'
%{_javadir}
jenkins@64f58e44d47d:/srv/releases/jenkins$ rpm --showrc | grep java
  %{?_javaclasspath:CLASSPATH="%{_javaclasspath}"
-13: __docdir_path	%{_datadir}/doc:%{_datadir}/man:%{_datadir}/info:%{_datadir}/gtk-doc/html:%{_datadir}/gnome/help:%{?_docdir}:%{?_mandir}:%{?_infodir}:%{?_javadocdir}:/usr/doc:/usr/man:/usr/info:/usr/X11R6/man
jenkins@64f58e44d47d:/srv/releases/jenkins$  rpm --version
RPM version 4.17.0
```

**Comment by timja** [comment]
Macro removed in https://github.com/rpm-software-management/rpm/commit/a9199f17be42d22791caa33c1d79157b29b811d2

**Comment by timja** [comment]
I think https://github.com/jenkins-infra/docker-packaging/pull/34 is the way to go

File: rpm/build/SPECS/jenkins.spec [hunk 1]
```diff
@@ -24,7 +24,7 @@ BuildRoot:	%{_tmppath}/build-%{name}-%{version}
 # TODO: If re-enable, fix the matcher for Java 11
 # Requires: java >= 1:1.8.0
 Requires: procps
-PreReq: /usr/sbin/groupadd /usr/sbin/useradd
```
Comment (timja) [inline rpm/build/SPECS/jenkins.spec:27]: this was deprecated

File: suse/build/SPECS/jenkins.spec [hunk 2]
```diff
@@ -25,7 +25,7 @@ BuildRoot:	%{_tmppath}/build-%{name}-%{version}
 # TODO: Fix the query for Java 11 if it is reenabled
 # Requires: java >= 1:1.8.0
 Requires:	procps
-PreReq:		/usr/sbin/groupadd /usr/sbin/useradd
```
Comment (timja) [inline suse/build/SPECS/jenkins.spec:28]: this was deprecated

File: suse/build/SPECS/jenkins.spec [hunk 3]
```diff
@@ -71,10 +71,10 @@ rm -rf "%{buildroot}"
 %__install -D -m0755 "%{SOURCE7}" "%{buildroot}%{_datadir}/%{name}/migrate"
 
 %pre
-/usr/sbin/groupadd -r %{name} &>/dev/null || :
+/usr/bin/getent group %{name} &>/dev/null || /usr/sbin/groupadd -r %{name} &>/dev/null
```
Comment (timja) [inline suse/build/SPECS/jenkins.spec:74]: this is the recommended method for adding a group depending on if it exists or not

File: suse/build/SPECS/jenkins.spec [hunk 4]
```diff
@@ -71,10 +71,10 @@ rm -rf "%{buildroot}"
...
 # SUSE version had -o here, but in Fedora -o isn't allowed without -u
-/usr/sbin/useradd -g %{name} -s /bin/false -r -c "@@SUMMARY@@" \
-	-d "%{workdir}" %{name} &>/dev/null || :
+/usr/bin/getent passwd %{name} &>/dev/null || /usr/sbin/useradd -g %{name} -s /bin/false -r -c "@@SUMMARY@@" \
```
Comment (timja) [inline suse/build/SPECS/jenkins.spec:76]: this is the recommended method for adding a user depending on if it exists or not

File: rpm/publish/publish.sh [hunk 5]
```diff
@@ -42,10 +42,10 @@ EOF
 	# generate index
 	# locally
 	# disable this for now, as it's currently now used and generate errors
-	# createrepo --update -o "$RPM_WEBDIR" "$RPMDIR/"
+	# createrepo_c --update -o "$RPM_WEBDIR" "$RPMDIR/"
 	# on the server
 	# shellcheck disable=SC2029
-	ssh "${SSH_OPTS[@]}" "$PKGSERVER" createrepo --update -o "'$RPM_WEBDIR'" "'$RPMDIR/'"
+	ssh "${SSH_OPTS[@]}" "$PKGSERVER" createrepo_c --update -o "'$RPM_WEBDIR'" "'$RPMDIR/'"
```
Comment (timja) [inline rpm/publish/publish.sh:48]: https://github.com/jenkins-infra/release/issues/202

**Comment by timja** [comment]
nice work on the test framework @basil!

The tests caught 2 issues (since fixed):
1. Building on Ubuntu no longer works for debian by default as it uses a compression algorithm not supported on debian
2. `rpmsign` uses gpgv2 now which wasn't installed and silently failed, tests caught that it wasn't signed anymore

File: rpm/publish/publish.sh (hunk 5 above)
Comment (basil) [reply rpm/publish/publish.sh:48]: It does not seem that `$PKGSERVER` has `createrepo_c` installed, which is a prerequisite to the successful deployment of this change.

Comment (basil) [reply rpm/publish/publish.sh:48]: > If no, could we use a Docker container for this instead?

It is just software. You can do anything you want, if you write the code.

In the short term, I suggest upgrading the box to Ubuntu 22.04.

Comment (dduportal) [reply rpm/publish/publish.sh:48]: > In the short term, I suggest upgrading the box to Ubuntu 22.04.

This machine also hosts the Update Center: its not an easy task and would require unplanned effort for now.
Not that I would not want, but that would be a huge context switch for the team meaning we would have to plan it for next week.

Comment (timja) [reply rpm/publish/publish.sh:48]: We can just switch back to createrepo and symlink it in the container, it’s API compatible as far as I understand

Comment (dduportal) [reply rpm/publish/publish.sh:48]: make sense. Many thanks @timja @basil for the help and quick reactions!

Upgrading our boxes from 18.04 to 22.04 is a topic that I would like to see happen quite fast though!

Comment (timja) [reply rpm/publish/publish.sh:48]: Err, I don't think createrepo is even needed in the docker container?

It is only executed over SSH?

Comment (timja) [reply rpm/publish/publish.sh:48]: https://release.ci.jenkins.io/job/core/job/package/job/master/200/console AND https://github.com/jenkinsci/packaging/pull/318

Comment (dduportal) [reply rpm/publish/publish.sh:48]: https://github.com/jenkins-infra/helpdesk/issues/2982 for info

**Pull request merged**